File: EOG-RNN-FeatureExtraction/src/feature_extraction.py
import numpy as np
import pandas as pd
from scipy.fft import fft
from scipy.stats import entropy, kurtosis, skew

def extract_time_domain_features(signal):
    features = {
        'signal_variance': np.var(signal),
        'signal_iqr': np.percentile(signal, 75) - np.percentile(signal, 25),
        'mean': np.mean(signal),
        'std': np.std(signal),
        'max': np.max(signal),
        'min': np.min(signal),
        'skew': skew(signal),
        'kurtosis': kurtosis(signal),
        'energy': np.sum(np.square(signal)),
        'median': np.median(signal),
        'entropy': entropy(np.abs(signal))
    }

    # Calculate the slope of the signal
    if len(signal) > 1:
        signal_slope = np.diff(signal)  # First derivative (slope)
        features['mean_slope'] = np.mean(signal_slope)
        features['std_slope'] = np.std(signal_slope)
    else:
        features['mean_slope'] = 0
        features['std_slope'] = 0

        
    return pd.Series(features)

def extract_frequency_domain_features(signal):
    fft_vals = fft(signal)
    fft_magnitude = np.abs(fft_vals)
    power = np.square(fft_magnitude)

    features = {
        'fft_max_freq': np.argmax(fft_magnitude),
        'fft_mean_power': np.mean(power),
        'fft_peak_power': np.max(power)
    }
    return pd.Series(features)

def extract_features(hEOG, vEOG):
    # Time-domain features
    hEOG_time_features = hEOG.apply(extract_time_domain_features, axis=1)
    hEOG_time_features.columns = ['h_' + col for col in hEOG_time_features.columns]
    vEOG_time_features = vEOG.apply(extract_time_domain_features, axis=1)
    vEOG_time_features.columns = ['v_' + col for col in vEOG_time_features.columns]

    # Frequency-domain features
    hEOG_freq_features = hEOG.apply(lambda row: extract_frequency_domain_features(row.values), axis=1)
    hEOG_freq_features.columns = ['h_' + col for col in hEOG_freq_features.columns]
    vEOG_freq_features = vEOG.apply(lambda row: extract_frequency_domain_features(row.values), axis=1)
    vEOG_freq_features.columns = ['v_' + col for col in vEOG_freq_features.columns]

    # Combine features
    combined_time_features = pd.concat([hEOG_time_features, vEOG_time_features], axis=1)
    combined_freq_features = pd.concat([hEOG_freq_features, vEOG_freq_features], axis=1)

    combined_features = pd.concat([combined_time_features, combined_freq_features], axis=1)

    return combined_features

File: EOG-RNN-FeatureExtraction/matlab/feature_extraction.m
function combined_features = feature_extraction(hEOG, vEOG)
    % Time-domain features
    hEOG_time_features = arrayfun(@(row) extract_time_domain_features(hEOG(row, :)), 1:size(hEOG, 1), 'UniformOutput', false);
    vEOG_time_features = arrayfun(@(row) extract_time_domain_features(vEOG(row, :)), 1:size(vEOG, 1), 'UniformOutput', false);

    hEOG_time_features = cell2mat(hEOG_time_features);
    vEOG_time_features = cell2mat(vEOG_time_features);

    % Frequency-domain features
    hEOG_freq_features = arrayfun(@(row) extract_frequency_domain_features(hEOG(row, :)), 1:size(hEOG, 1), 'UniformOutput', false);
    vEOG_freq_features = arrayfun(@(row) extract_frequency_domain_features(vEOG(row, :)), 1:size(vEOG, 1), 'UniformOutput', false);

    hEOG_freq_features = cell2mat(hEOG_freq_features);
    vEOG_freq_features = cell2mat(vEOG_freq_features);

    % Combine features
    combined_features = [hEOG_time_features, vEOG_time_features, hEOG_freq_features, vEOG_freq_features];
end

function features = extract_time_domain_features(signal)
    features.signal_variance = var(signal);
    features.signal_iqr = iqr(signal);
    features.mean = mean(signal);
    features.std = std(signal);
    features.max = max(signal);
    features.min = min(signal);
    features.skew = skewness(signal);
    features.kurtosis = kurtosis(signal);
    features.energy = sum(signal.^2);
    features.median = median(signal);
    features.entropy = entropy(signal);  % You need to implement or find an entropy function

    % Calculate the slope of the signal
    if length(signal) > 1
        signal_slope = diff(signal);  % First derivative (slope)
        features.mean_slope = mean(signal_slope);
        features.std_slope = std(signal_slope);
    else
        features.mean_slope = 0;
        features.std_slope = 0;
    end
end

function features = extract_frequency_domain_features(signal)
    fft_vals = fft(signal);
    fft_magnitude = abs(fft_vals);
    power = fft_magnitude.^2;

    features.fft_max_freq = find(fft_magnitude == max(fft_magnitude)) - 1;  % MATLAB index starts from 1
    features.fft_mean_power = mean(power);
    features.fft_peak_power = max(power);
end
